### ISSUE TITLE
DALIGenericIterator repeat first batch as output.

### ISSUE THREAD

**Issue by Approximetal** [NVIDIA/DALI#2018]
Hi, there are two issues when I run the training model with pytorch. 
1. I found the outputs of `enumerate(dali_iter)` are the same in each iteration. 
2. When I run my code on a large dataset, the program will get stuck somewhere after thousands of iterations(it usually happens after few hours, the duration is random) without error and warning, but it won't happen when I test it on a small dataset(may because the test not long enough).

I was wondering if you know the reason. I list part of my output below. It can be seen the dataset size and value are the same in different iterations and epochs.

```
dataset size:724, batch size:4
epoch 0, iteration 0
mel_inputs: torch.Size([4, 80, 362])
input_lengths: [tensor([362.]), tensor([220.]), tensor([170.]), tensor([133.])]
mel_targets: torch.Size([4, 80, 327])
output_lengths: [tensor([327.]), tensor([198.]), tensor([196.]), tensor([144.])]
speaker_id: [tensor([2101., 2115.]), tensor([2101., 2088.]), tensor([2101., 2086.]), tensor([2101., 2087.])]
epoch 0, iteration 10
mel_inputs: torch.Size([4, 80, 362])
input_lengths: [tensor([362.]), tensor([220.]), tensor([170.]), tensor([133.])]
mel_targets: torch.Size([4, 80, 327])
output_lengths: [tensor([327.]), tensor([198.]), tensor([196.]), tensor([144.])]
speaker_id: [tensor([2101., 2115.]), tensor([2101., 2088.]), tensor([2101., 2086.]), tensor([2101., 2087.])]
......
epoch 0, iteration 180
mel_inputs: torch.Size([4, 80, 362])
input_lengths: [tensor([362.]), tensor([220.]), tensor([170.]), tensor([133.])]
mel_targets: torch.Size([4, 80, 327])
output_lengths: [tensor([327.]), tensor([198.]), tensor([196.]), tensor([144.])]
speaker_id: [tensor([2101., 2115.]), tensor([2101., 2088.]), tensor([2101., 2086.]), tensor([2101., 2087.])]
WARNING:root:DALI iterator does not support resetting while epoch is not finished. Ignoring...

epoch 1, iteration 0
mel_inputs: torch.Size([4, 80, 362])
input_lengths: [tensor([362.]), tensor([220.]), tensor([170.]), tensor([133.])]
mel_targets: torch.Size([4, 80, 327])
output_lengths: [tensor([327.]), tensor([198.]), tensor([196.]), tensor([144.])]
speaker_id: [tensor([2101., 2115.]), tensor([2101., 2088.]), tensor([2101., 2086.]), tensor([2101., 2087.])]
epoch 1, iteration 10
mel_inputs: torch.Size([4, 80, 362])
input_lengths: [tensor([362.]), tensor([220.]), tensor([170.]), tensor([133.])]
mel_targets: torch.Size([4, 80, 327])
output_lengths: [tensor([327.]), tensor([198.]), tensor([196.]), tensor([144.])]
speaker_id: [tensor([2101., 2115.]), tensor([2101., 2088.]), tensor([2101., 2086.]), tensor([2101., 2087.])]
epoch 1, iteration 20
mel_inputs: torch.Size([4, 80, 362])
input_lengths: [tensor([362.]), tensor([220.]), tensor([170.]), tensor([133.])]
mel_targets: torch.Size([4, 80, 327])
output_lengths: [tensor([327.]), tensor([198.]), tensor([196.]), tensor([144.])]
speaker_id: [tensor([2101., 2115.]), tensor([2101., 2088.]), tensor([2101., 2086.]), tensor([2101., 2087.])]
......
```

**Comment by JanuszL** [comment]
Hi,

> I found the outputs of enumerate(dali_iter) are the same in each iteration.

Have you set `random_shuffle=True` in the reader you use in the pipeline?

> When I run my code on a large dataset, the program will get stuck somewhere after thousands of iterations(it usually happens after few hours, the duration is random) without error and warning, but it won't happen when I test it on a small dataset(may because the test not long enough).

It is not expected. Does DALI just stop and you cannot get the output from it or the whole training gets stuck? How the CPU and GPU utilization looks like (I wonder if this is some infinite loop or some deadlock)?
Can you provide a pipeline you are using?

**Comment by Approximetal** [comment]
> Have you set random_shuffle=True in the reader you use in the pipeline?

I use ExternalSource following this [example](https://github.com/NVIDIA/DALI/blob/cde7271a840142221273f8642952087acd919b6e/docs/examples/frameworks/pytorch/pytorch-external_input.ipynb) I have set the shuffle in `ExternalInputIterator`, code is like:
```
    def __iter__(self):
        self.i = 0
        shuffle(self.filepaths_pair)
        return self
```
 
> Does DALI just stop and you cannot get the output from it or the whole training gets stuck? How the CPU and GPU utilization looks like (I wonder if this is some infinite loop or some deadlock)?

The weirdest thing is the program is running, CPU and GPU utilization looks normal, like everything is running, all going well except no log output, and tensorboard stop update. The pipeline code can be found in [here](https://github.com/NVIDIA/DALI/issues/2007#issue-634520448)

**Comment by JanuszL** [comment]
It looks like `shuffle` returns always the same order every-time the iterator is created. Please check `shuffle` API for more details.

The pipeline looks good, What you can do it add print inside the `iter_setup` and at the output from the pipeline to see if you are getting data from the pipeline. In the case that you have not provided the input properly to the DALI pipeline inside the `iter_setup` DALI pipeline will hang waiting for it.
To mitigate this kind of problem I would recommend using new [`ExternalSource`](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.ops.ExternalSource) API and pass the iterator directly to the operator, like in [this example](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/data_loading/external_input.html).

**Comment by Approximetal** [comment]
I added the print inside the `iter_setup`, it printed the output in each iteration. 

> I would recommend using new ExternalSource API and pass the iterator directly to the operator, like in this example.

What is the differences between the example you recommend, with the one I referenced? Do you mean I should add some parameters in  `nvidia.dali.ops.ExternalSource()` function?

**Comment by JanuszL** [comment]
>I added the print inside the iter_setup, it printed the output in each iteration.

And do you see the print from it before hang or it hangs before that print?

In the example I have provided:
```
 self.source = ops.ExternalSource(source = eii, num_outputs = 2)
```
Iterator is passed directly to the `ExternalSource` so no more error-prone `iter_setup`, and one `ExternalSource` can provide multiple output tensors. The example you use is outdated.

**Comment by Approximetal** [comment]
The program hasn't hang as I just add `print` in `iter_setup` in a moment. I will update the result tomorrow.

> ` self.source = ops.ExternalSource(source = eii, num_outputs = 2)`

I follow the new example and get this error:
```
  File "/home/zzy/anaconda3/envs/StarGAN-VC/lib/python3.6/site-packages/nvidia/dali/ops.py", line 476, in __init__
    dtype = self._schema.GetArgumentType(key)
RuntimeError: [/opt/dali/dali/pipeline/operator/op_schema.cc:127] Assert on "HasArgument(name)" failed: Argument "source" is not supported by operator "ExternalSource".
```
How to deal with it?

**Comment by JanuszL** [comment]
I think you need to update you DALI version to the most recent one.

**Comment by Approximetal** [comment]
I just update the DALI version to `nvidia-dali            0.22.0`,  and get the same error.

**Comment by JanuszL** [comment]
Just to be sure, you are using the `source` argument at the `__init__` stage, not inside the `define_graph` method, right?
@mzient and clue what is wrong there?

**Comment by Approximetal** [comment]
Yes. I changed my code like this:
```
class ExternalSourcePipeline(Pipeline):
    def __init__(self, batch_size, num_threads, device_id,  external_data):
        super(ExternalSourcePipeline, self).__init__(batch_size, num_threads, device_id, seed=12)
        self.source = ops.ExternalSource(source = external_data, num_outputs = 8)
        self.pad = ops.Pad(fill_value=0)
        self.gate_pad = ops.Pad(fill_value=1)
```

**Comment by mzient** [comment]
It looks like the update didn't work and you're still with old DALI. For once, your error message refers to line ops.py:476 as containing `dtype = self._schema.GetArgumentType(key)`, but it's not what should be in this line.

**Comment by JanuszL** [comment]
Can you run pip show `nvidia-dali`?

**Comment by Approximetal** [comment]
$ pip show nvidia-dali
Name: nvidia-dali
Version: 0.22.0
Summary: NVIDIA DALI for CUDA 10.0.130. Git SHA: 4070053b806489922e6f1bf9b3ba74328468a884
Home-page: https://github.com/NVIDIA/dali
Author: NVIDIA Corporation
Author-email: UNKNOWN
License: Apache License 2.0
Location: /home/zzy/.local/lib/python3.7/site-packages
Requires: future
Required-by:

**Comment by Approximetal** [comment]
I found the reason of first issue. In `ExternalInputIterator` I use `i` rather than `self.i` to select input variables. So I changed the code from:
```
        for i in range(self.batch_size):
            file = self.get_mel_pair(self.filepaths_pair[i])
```
to
```
        for _ in range(self.batch_size):
            file = self.get_mel_pair(self.filepaths_pair[self.i])
```
But it brings another issue:
```
    for i, data in enumerate(dali_iter):
  File "/home/zzy/anaconda3/envs/StarGAN-VC/lib/python3.6/site-packages/nvidia/dali/plugin/pytorch.py", line 224, in __next__
    feed_ndarray(tensor, pyt_tensors[category])
  File "/home/zzy/anaconda3/envs/StarGAN-VC/lib/python3.6/site-packages/nvidia/dali/plugin/pytorch.py", line 53, in feed_ndarray
    ", but PyTorch Tensor has size {1}".format(dali_tensor.shape(), list(arr.size())))
AssertionError: Shapes do not match: DALI tensor has size [4, 80, 447], but PyTorch Tensor has size [4, 80, 492]
```
I suppose it is as the same as [this reason](https://github.com/NVIDIA/DALI/issues/2007#issuecomment-640577492)
Sadly I cannot run it now.

**Comment by JanuszL** [comment]
Hi,
Please set `dynamic_shape` in [the DALI iterator](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/plugins/pytorch_plugin_api.html) and see if that helps?

**Comment by Approximetal** [comment]
> And do you see the print from it before hang or it hangs before that print?

It hangs before the print in `iter_setup`.

> Please set dynamic_shape in the DALI iterator and see if that helps?

It works, but the speed of DALI dataloader is even slower than before.

**Comment by JanuszL** [comment]
> It works, but the speed of DALI dataloader is even slower than before.

It makes DALI reallocate PyTorch Tensor whenever shape changes, so it may be slower a bit.

> I stopped the task manually, I was wondering if it is the position that the error shows is the place the task hangs. If so, it should be a bug in my code.

It doesn't as it hangs on the native side. Without repro it is hard to tell more what may have happened.

**Comment by Approximetal** [comment]
Thanks for the reply, hope the team can optimize DALI to speed up variable-length input in the future. XD

**Comment by JanuszL** [comment]
When you use `dynamic_shape` DL FWs tensors are keep recreated and there is extra zeroing step which may slow things down. I added a PR that may help https://github.com/NVIDIA/DALI/pull/2030, but it needs more testing.
